src/coprocessor/circom.rs

```diff
@@ -0,0 +1,159 @@
+        );
+    }
+
+    fn validate_gadget<F: LurkField, C: CircomGadget<F>>(gadget: &C) -> Result<()> {
+        if !circom_dir().exists() {
+            std::fs::create_dir_all(circom_dir())?;
+            return print_error(gadget.name(), vec![]);
+        }
+
+        let name = gadget.name();
+        let circom_folder = circom_dir().join(name);
+
+        if circom_folder.exists() {
+            return Ok(());
+        };
+
+        let mut subdirs = Vec::new();
+
+        for entry in read_dir(circom_dir())? {
+            let entry = entry?;
+            let path = entry.path();
+
+            if path.is_dir() {
+                if let Some(dir_name) = path.file_name() {
+                    if let Some(dir_name) = dir_name.to_str() {
+                        subdirs.push(dir_name.to_string());
+                    }
+                }
+            }
+        }
+
+        if subdirs.contains(&gadget.name().into()) {
+            return Ok(());
+        }
+
+        print_error(gadget.name(), subdirs)
+    }
+
+    #[derive(Debug)]
+    pub struct CircomCoprocessor<F: LurkField, C: CircomGadget<F>> {
+        gadget: C,
+        config: CircomConfig<F>,
+    }
+
+    impl<F: LurkField, C: CircomGadget<F>> Clone for CircomCoprocessor<F, C> {
+        fn clone(&self) -> Self {
+            CircomCoprocessor::new(self.gadget.clone())
+        }
+    }
+
+    impl<F: LurkField, C: CircomGadget<F>> CoCircuit<F> for CircomCoprocessor<F, C> {
+        /// TODO: Generalize
+        fn arity(&self) -> usize {
+            0
+        }
+
+        fn synthesize<CS: ConstraintSystem<F>>(
+            &self,
+            cs: &mut CS,
+            g: &GlobalAllocations<F>,
+            _store: &Store<F>,
+            input_exprs: &[AllocatedPtr<F>],
+            input_env: &AllocatedPtr<F>,
+            input_cont: &AllocatedContPtr<F>,
+        ) -> Result<(AllocatedPtr<F>, AllocatedPtr<F>, AllocatedContPtr<F>), SynthesisError>
+        {
+            let input = self.gadget.clone().into_circom_input(input_exprs);
+            let witness = circom_scotia::calculate_witness(&self.config, input, true).expect("msg");
+            let output = circom_scotia::synthesize(cs, self.config.r1cs.clone(), Some(witness))?;
+
+            let res = AllocatedPtr::from_parts(g.num_tag.clone(), output);
+
+            Ok((res, input_env.clone(), input_cont.clone()))
+        }
+    }
+
+    impl<F: LurkField, C: CircomGadget<F> + Debug> Coprocessor<F> for CircomCoprocessor<F, C> {
+        /// TODO: Generalize
+        fn eval_arity(&self) -> usize {
+            0
+        }
+
+        fn simple_evaluate(&self, s: &mut Store<F>, args: &[Ptr<F>]) -> Ptr<F> {
+            self.gadget.simple_evaluate(s, args)
+        }
+
+        fn has_circuit(&self) -> bool {
+            true
+        }
+    }
+
+    impl<F: LurkField, C: CircomGadget<F>> CircomCoprocessor<F, C> {
+        /// Creates a CircomConfig by loading in the data in `.lurk/circom-gadgets/<name>/*`
```

Comment:
Is this PR to be interpreted under the assumption some gadgets are available locally, but are not part of this PR?